src/components/feedback/Dialog.tsx

```diff
@@ -53,46 +66,42 @@ const DialogNext = function Dialog({
   children,
   initialFocus = 'auto',
   restoreFocus = false,
+  transitionComponent: TransitionComponent,
 
   classes,
   elementRef,
 
   // Forwarded to Panel
   buttons,
   icon,
-  onClose,
+  onClose = noop,
   paddingSize = 'md',
   scrollable = true,
   title,
 
   ...htmlAttributes
 }: DialogProps) {
   const modalRef = useSyncedRef(elementRef);
-  const closeHandler = onClose ?? noop;
   const restoreFocusEl = useRef<HTMLElement | null>(
     document.activeElement as HTMLElement | null
   );
+  const [transitionComponentVisible, setTransitionComponentVisible] =
+    useState(false);
 
-  useClickAway(modalRef, closeHandler, {
-    enabled: closeOnClickAway,
-  });
-
-  useKeyPress(['Escape'], closeHandler, {
-    enabled: closeOnEscape,
-  });
-
-  useFocusAway(modalRef, closeHandler, {
-    enabled: closeOnFocusAway,
-  });
-
-  const dialogDescriptionId = useUniqueId('dialog-description');
+  const closeHandler = useCallback(() => {
+    if (TransitionComponent) {
+      setTransitionComponentVisible(false);
+    } else {
+      onClose();
+    }
+  }, [onClose, TransitionComponent]);
```

Comment:
If you felt like it added useful context, you _could_ add a brief comment or two in this handler function.

Reply:
I think I'll just a short note on the first if, explaining "why" this is done like this.